spec/scheduling-tasks.md

```diff
@@ -0,0 +1,380 @@
+
+<div algorithm="schedule a postTask task">
+  To <dfn lt="schedule a postTask task|scheduling a postTask task">schedule a postTask task</dfn>
+  for {{Scheduler}} |scheduler| given a {{SchedulerPostTaskCallback}} |callback| and
+  {{SchedulerPostTaskOptions}} options, run the following steps:
+
+  1. Let |result| be [=a new promise=].
+  1. Let |signal| be |options|["{{SchedulerPostTaskOptions/signal}}"].
+  1. If |signal| is not null and its [=AbortSignal/aborted flag=] is set, then
+     [=reject=] |result| with an "{{AbortError!!exception}}" {{DOMException}}
+     and return |result|.
+  1. Let |priority| be |options|["{{SchedulerPostTaskOptions/priority}}"].
+  1. Let |queue| be the result of [=selecting the scheduler task queue=] for
+     |scheduler| given |signal| and |priority|.
+  1. Let |delay| be |options|["{{SchedulerPostTaskOptions/delay}}"].
+  1. If |delay| is less than 0 then set |delay| to 0.
+  1. If |delay| is greater than 0, then the task is a delayed task; return
+     |result| and run the following steps [=in parallel=]:
+    1. Let |global| be the [=relevant global object=] for |scheduler|.
+    1. If |global| is a {{Window}} object, wait until |global|'s
+       <a attribute for="Window">associated <code>Document</code></a>
+       has been fully active for a further |delay| milliseconds (not necessarily
+       consecutively).
+
+       Otherwise, |global| is a {{WorkerGlobalScope}} object; wait until |delay|
+       milliseconds have passed with the worker not suspended (not necessarily
+       consecutively).
+
+    1. Wait until any invocations of this algorithm that had the same |scheduler|,
+       that started before this one, and whose |delay| is equal to or less
+       than this one's, have completed.
+    1. Optionally, wait a further [=implementation-defined=] length of time.
+    1. [=Schedule a task to invoke a callback=] for |scheduler| given |queue|,
+       |signal|, |callback|, and |result|.
+  1. Otherwise the task is not delayed. [=Schedule a task to invoke a callback=]
+     for |scheduler| given |queue|, |signal|, |callback|, and |result|.
+</div>
+
+Issue: We need to figure out exactly how we want to spec delayed tasks, and if
+we can refactor the timer spec to use a common method. As written, this uses
+steps 15&ndash;17 of the timer initialization steps algorithm, but there are a
+couple things we might want to change: (1) how to account for suspend? (2) how
+to account for current throttling techniques (see also
+[this issue](https://github.com/whatwg/html/issues/5925))?
+
+<div algorithm="select the scheduler task queue">
+  To <dfn lt="select the scheduler task queue|selecting the scheduler task queue">select the scheduler task queue</dfn>
+  for a {{Scheduler}} |scheduler| given ({{TaskSignal}} or {{AbortSignal}})
+  |signal|, and a {{TaskPriority}} |priority|:
+
+  1. If |priority| is null and |signal| is not null and |signal| is a
```

Comment:
You can be extra-precise by using `|signal| [=implements=] the {{TaskSignal}} interface`